flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateWithExecutionGraph.java

```diff
@@ -306,22 +327,87 @@ void deliverOperatorEventToCoordinator(
+            FailureResult failureResult, T context) {
+        if (failureResult.canRestart()) {
+            getLogger().info("Restarting job.", failureResult.getFailureCause());
+            context.goToRestarting(
+                    getExecutionGraph(),
+                    getExecutionGraphHandler(),
+                    getOperatorCoordinatorHandler(),
+                    failureResult.getBackoffTime(),
+                    getFailures());
+        } else {
+            getLogger().info("Failing job.", failureResult.getFailureCause());
+            context.goToFailing(
+                    getExecutionGraph(),
+                    getExecutionGraphHandler(),
+                    getOperatorCoordinatorHandler(),
+                    failureResult.getFailureCause(),
+                    getFailures());
+        }
+    }
+
+    /**
+     * Transition to different state when the execution graph reaches a globally terminal state.
+     *
+     * @param globallyTerminalState globally terminal state which the execution graph reached
+     */
+    abstract void onGloballyTerminalState(JobStatus globallyTerminalState);
+
+    @Override
+    public void handleGlobalFailure(Throwable cause) {
+        failureCollection.add(ExceptionHistoryEntry.createGlobal(cause));
+        onFailure(cause);
+    }
+
     /**
      * Updates the execution graph with the given task execution state transition.
      *
      * @param taskExecutionStateTransition taskExecutionStateTransition to update the ExecutionGraph
      *     with
      * @return {@code true} if the update was successful; otherwise {@code false}
      */
-    abstract boolean updateTaskExecutionState(
-            TaskExecutionStateTransition taskExecutionStateTransition);
+    boolean updateTaskExecutionState(TaskExecutionStateTransition taskExecutionStateTransition) {
+        // collect before updateState, as updateState may deregister the execution
+        final Optional<AccessExecution> maybeExecution =
+                executionGraph.findExecution(taskExecutionStateTransition.getID());
+        final Optional<String> maybeTaskName =
+                executionGraph.findVertexWithAttempt(taskExecutionStateTransition.getID());
+
+        final ExecutionState desiredState = taskExecutionStateTransition.getExecutionState();
+        boolean successfulUpdate = getExecutionGraph().updateState(taskExecutionStateTransition);
+        if (successfulUpdate && desiredState == ExecutionState.FAILED) {
+            final AccessExecution execution =
+                    maybeExecution.orElseThrow(NoSuchElementException::new);
+            final String taskName = maybeTaskName.orElseThrow(NoSuchElementException::new);
+            final ExecutionState currentState = execution.getState();
+            if (currentState == desiredState) {
+                failureCollection.add(ExceptionHistoryEntry.create(execution, taskName));
```

Comment:
I do wonder whether this collection should be bounded. The exception history as a whole is, and this here not having a bound seems counter to that idea.

Reply:
The failureCollection contains all exceptions that were thrown in the interval from the initial failure to the job restart or cancellation. Usually this should be no more than one exception. Local failures are bounded by the number of tasks (actually the number is currently limited to 1). I don't know whether/how the number of global failures is bounded, but all failures would have to arrive in this short time interval.  

Given these restrictions, I don't think that having this collection unbounded is really dangerous. 

If you want to change it anyway, should we add a new config parameter for this? Were should we add this, also to the `WebOptions`?

Reply:
> Usually this should be no more than one exception

That's a big assumption. If a TM dies that is part of a shuffle you could easily and up with with `<parallelism>` number of exceptions.

But the DefaultScheduler also doesn't limit the exceptions, so it should be fine.